woocommerce/woocommerce-ios
# Order Creation: Refactor address form, prepare for reusability

Part of https://github.com/woocommerce/woocommerce-ios/issues/5412.

## Description

This PR adds support for multiple nested forms to address flow UI and updates country/state values handling in address flow.

The plan is to consolidate all address data in `fields` and add second property similar to existing
```
var fields: AddressFormFields
```
This will likely require a few additional properties, like a second pair of country/state view models constructors. I aim to keep them in `CreateOrderAddressFormViewModel`, but some (at least second `fields`) will have to be added on top level (`AddressFormViewModelProtocol`).

I have a working draft of the next step, but if you have any ideas for architecture improvements in this scope - please share!

## Changes

- Adds `differentAddressToggle` support to address form.
- Adds `SingleAddressForm` to allow multiple reusable nested forms in address flow.
- Adds second address input and output arguments for creation flow (unused now).
- Updates country and state values flow to keep all values in `fields` and simplify update flow.
- Replaces `GeometryReader` with `safeAreaInsets` from Environment.

### Known limitations

The second address form is unused, just a placeholder now. First address data will be duplicated there.

## Testing

There is not much new functionality added, focus on existing new order and edit order flows.

Setup:
1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.

Try new order address flow:
1. Go to the Orders tab and tap the + button. Tap "Create order" in the bottom sheet.
2. Tap "+ Add customer" button to display the address form.
3. Enter some data (pay attention to Country and State).
4. Tap "Done" to save and close the address form.
5. Confirm that the customer section is updated with new data.
6. Tap "Edit" in the customer section.
7. Confirm that address form. shows correct data.
8. Clear all fields and tap "Done".
9. Confirm that the empty state for both addresses shows "No address specified.".
10. Tap "Create" to create an order. Confirm that the order displays the correct address.

Try existing order editing flow:
11. In order details tap the pencil icon to edit the shipping address.
12. Update some data (pay attention to Country and State).
13. Tap "Done" to save and close the address form.
14. Confirm that the customer section is updated with new data.

---
